README.md
# steven_challenege
TLDR: Look for the SOLUTION.md in each project for the solution to each coding question

--------------------------------------------------------------------------------------------------------------

## Infrastructure
For this project, please think about how you would architect a scalable and secure static web application in AWS.
    • Create and deploy a running instance of a web server using a configuration management tool of your choice. The web server should serve one page with the following content.<br>
```
<html>
<head>
<title>Hello World</title>
</head>
<body>
<h1>Hello World!</h1>
</body>
</html>
```

    • Secure this application and host such that only appropriate ports are publicly exposed and any http requests are redirected to https. This should be automated using a configuration management tool of your choice and you should feel free to use a self-signed certificate for the web server.
    • Develop and apply automated tests to validate the correctness of the server configuration.
    • Express everything in code.
    • Provide your code in a Git repository named <FIRSTNAME>_Challenge on GitHub.com
Be prepared to walk though your code, discuss your thought process, and talk through how you might monitor and scale this application. You should also be able to demo a running instance of the host.

## Coding Challenge
```
Please solve the below problem in python or go. 
You don't need to do the submission on the web site. Include the program in the repo above.
    . https://www.hackerrank.com/challenges/validating-credit-card-number/problem
```



ChangeLog
Correct minor build issues with various bucket ACLs

commit ab572f51c716f3904438586883e83ca24f9d07f9
Author: Steven Edwards <winehacker@gmail.com>
Date:   Wed Jun 28 08:35:37 2023 -0700

    - Started implementing IAM Policy for s3 bucket to access CloudFront
    - Start working on cleaning up the CC number tests

commit 1ef617e04c557d4bd695da2551bc17fa335b3b2a
Author: Steven Edwards <winehacker@gmail.com>
Date:   Wed Jun 28 06:18:37 2023 -0700

    Started Implementing CloudFront

commit 0b70d99309713a5f5cbf28a94a082599d3a033fa
Author: Steven Edwards <winehacker@gmail.com>
Date:   Wed Jun 28 03:00:16 2023 -0700

    Started Working on Certifcate Management for CloudFront

commit 59916dd0c3066ba2337a3aade3a4ae4a20ca9473
Author: Steven Edwards <winehacker@gmail.com>
Date:   Wed Jun 28 00:31:42 2023 -0700

    Correct track terraform state

commit 559b8d74039b3ea706459bfc79d4d17797341aa1
Author: Steven Edwards <winehacker@gmail.com>
Date:   Wed Jun 28 00:26:30 2023 -0700

    Encrypt bucket with correct key stored in KMS

commit bdf43fb0e428c38bfaaf8ba7c2440e879e70a96e
Author: Steven Edwards <winehacker@gmail.com>
Date:   Tue Jun 27 23:55:59 2023 -0700

    Checkpoint, start work on s3 bucket

commit e75ad721c3c9677857253585a598fbb64a84a11a
Author: Steven Edwards <winehacker@gmail.com>
Date:   Tue Jun 27 22:17:18 2023 -0700

    Expand out Go Credit Card Processor Example

commit 64f4e469ebc36b25511a901633cba6bf3ae19eed
Author: Steven Edwards <winehacker@gmail.com>
Date:   Tue Jun 27 21:35:14 2023 -0700

    Terraform Scaffolding for s3, certificates, etc

commit 121584cd245070a9555ea9e2a164d4e5532971c8
Author: Steven Edwards <winehacker@gmail.com>
Date:   Tue Jun 27 20:22:02 2023 -0700

    Document Challenge requirements

commit 56f060af519349b27ae96af88ea9edd4bcb8ae50
Author: sedwards <winehacker@gmail.com>
Date:   Tue Jun 27 19:02:21 2023 -0700

    Initial commit


coding/go_validator/scan_file.go
/*
 * Example of how to read line, by line. provided by
 * https://zetcode.com/golang/readfile/
 */

package main

import (
    "bufio"
    "fmt"
    "log"
    "os"
    "regexp"
)



func scan_file() {
    re := regexp.MustCompile(`^(?:4[0-9]{12}(?:[0-9]{3})?|[25][1-7][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})$`)

    fmt.Printf("Using Pattern for regex\n: %v\n", re.String()) // print pattern

    f, err := os.Open("./input/sample_input.txt")
    // o := os.Open("./output/go_sample_output.txt")
    if err != nil {
        log.Fatal(err)
    }

    defer f.Close()
    scanner := bufio.NewScanner(f)

    for scanner.Scan() {
        //fmt.Println(scanner.Text())
	//str cc_number := scanner.Text()
	cc_number := scanner.Text()
	fmt.Printf("CC : %v :%v\n", cc_number, re.MatchString(cc_number))
    }

    if err := scanner.Err(); err != nil {
        log.Fatal(err)
    }
}



coding/python_validator/python_validator.py
# Python3 program to validate Visa
# Card number using regular expression
import re
 
# Function to validate Visa Card
# number using regular expression.
def isValidVisaCardNo(string):
     
    # Regex to check valid Visa
    # Card number
    regex = "^4[0-9]{12}(?:[0-9]{3})?$";
 
    # Compile the ReGex
    p = re.compile(regex);
     
    # If the string is empty
    # return false
    if (string == ''):
        return False;
         
    # Pattern class contains matcher()
    # method to find matching between
    # given string and regular expression.
    m = re.match(p, string);
     
    # Return True if the string
    # matched the ReGex else False
    if m is None:
        return False
    else:
        return True
 
# Driver code
if __name__ == "__main__":
     
    # Test Case 1
    str1 = "4155279860457";
    print(str1, isValidVisaCardNo(str1));
     
    # Test Case 2
    str2 = "4155279860457201";
    print(str2, isValidVisaCardNo(str2));
     
    # Test Case 3
    str3 = "4155279";
    print(str3, isValidVisaCardNo(str3));
     
    # Test Case 4
    str4 = "6155279860457";
    print(str4, isValidVisaCardNo(str4));
     
    # Test Case 5
    str5 = "415a27##60457";
    print(str5, isValidVisaCardNo(str5));
     
# This code is contributed by AnkitRai01
# https://www.geeksforgeeks.org/how-to-validate-visa-card-number-using-regular-expression/



coding/go_validator/examples/example.go
/* Example credit to
   https://www.golangprograms.com/regular-expression-to-validate-common-credit-card-numbers.html

   Modified to support the example requirements of
   https://www.hackerrank.com/challenges/validating-credit-card-number/problem
*/

package main

import (
	"fmt"
	"regexp"
)

func main() {
	str1 := "4111111111111111"
	str2 := "346823285239073"
	str3 := "370750517718351"
	str4 := "4556229836495866"
	str5 := "5019717010103742"
	str6 := "76009244561"
	str7 := "4111-1111-1111-1111"
	str8 := "5610591081018250"
	str9 := "30569309025904"
	str10 := "6011111111111117"

	re := regexp.MustCompile(`^(?:4[0-9]{12}(?:[0-9]{3})?|[25][1-7][0-9]{14}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\d{3})\d{11})$`)

	fmt.Printf("Pattern: %v\n", re.String()) // print pattern
	fmt.Printf("\nCC : %v :%v\n", str1, re.MatchString(str1))
	fmt.Printf("CC : %v :%v\n", str2, re.MatchString(str2))
	fmt.Printf("CC : %v :%v\n", str3, re.MatchString(str3))
	fmt.Printf("CC : %v :%v\n", str4, re.MatchString(str4))
	fmt.Printf("CC : %v :%v\n", str5, re.MatchString(str5))
	fmt.Printf("CC : %v :%v\n", str6, re.MatchString(str6))
	fmt.Printf("CC : %v :%v\n", str7, re.MatchString(str7))
	fmt.Printf("CC : %v :%v\n", str8, re.MatchString(str8))
	fmt.Printf("CC : %v :%v\n", str9, re.MatchString(str9))
	fmt.Printf("CC : %v :%v\n", str10, re.MatchString(str10))
}

